### PR TITLE
UIDATIMP-486 MARCBib logic and validation for delete action

### PR DESCRIPTION
**Pull request by IvanKryzhanovskyi**
## Purpose
To specify the Delete details for the MARC record modifications field mapping profile

## Approach

- Use connect from 'react-redux' to fill correspond empty fields with default value
- Replace `validateAlphanumericValue` with `validateAlphanumericOrAllowedValue` in formValidators
- Add `fillFieldsWithDefaultValue` ro 'MARCTableRow' component
- Add validation rules
- Add tests.

## Ticket
[UIDATIMP-486](https://issues.folio.org/browse/UIDATIMP-486)

## Screenshot
fillEmptyFieldWithDefaultValue

### CODE REVIEW

File: src/utils/formValidators.js [hunk 1]
```diff
@@ -54,27 +54,32 @@ export const validateFileExtension = value => {
...
  * // => Translated string (en = 'Please enter an alphanumeric value')
  */
-export const validateAlphanumericValue = value => {
+export const validateAlphanumericOrAllowedValue = (value, allowedValue) => {
   const pattern = /^[a-zA-Z0-9]*$/;
   const val = value && value.trim ? value.trim() : value;
+  const isEmptyField = !val || !val.length;
 
-  if (!val || !val.length || val.match(pattern)) {
+  if (isEmptyField || val === allowedValue || val.match(pattern)) {
```
Comment (VictorSoroka1) [inline src/utils/formValidators.js:82]: I suggest using isEmpty util from lodash here.
```suggestion
  if (isEmpty(val) || val === allowedValue || val.match(pattern)) {
```

File: src/components/MARCTable/MARCTableRow.js [hunk 2]
```diff
@@ -608,4 +653,5 @@ MARCTableRow.defaultProps = {
   indicator1: '',
   indicator2: '',
   subfieldsData: [{}],
+  dispatchOnChange: () => {},
```
Comment (VictorSoroka1) [inline src/components/MARCTable/MARCTableRow.js:656]: I suggest using lodash noop util here.
```suggestion
  dispatchOnChange: noop,
```

File: src/components/MARCTable/MARCTableRow.js [hunk 3]
```diff
@@ -123,22 +158,31 @@ export const MARCTableRow = ({
...
 
       return null;
     },
     [actionValue, fieldValue],
   );
 
-  const handleActionChange = e => {
-    const { ADD } = MAPPING_DETAILS_ACTIONS;
+  const fillFieldsWithDefaultValue = (items, valueToFill) => items
+    .map(item => !item.value && dispatchOnChange(MAPPING_DETAILS_FORM_NAME, item.path, valueToFill));
```
Comment (VictorSoroka1) [inline src/components/MARCTable/MARCTableRow.js:174]: Could you please clarify what is the return value from `dispatchOnChange` is going to be?

File: src/components/MARCTable/MARCTableRow.js [hunk 4]
```diff
@@ -85,14 +98,28 @@ export const MARCTableRow = ({
...
   }, [indicator2]);
+  useEffect(() => {
+    setSubfieldValue(subfield);
+  }, [subfield]);
 
   const rowSubactions = allowedSubactions[actionValue] || [];
   const rowPositions = allowedPositions[actionValue] || {};
   const rowHasDataField = hasDataField[actionValue];
 
+  const fieldsToCheckArr = [{
```
Comment (VictorSoroka1) [inline src/components/MARCTable/MARCTableRow.js:109]:
```suggestion
  const fieldsToCheck = [{
```

File: src/components/MARCTable/MARCTableRow.js (hunk 2 above)
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: refactored

File: src/components/MARCTable/MARCTableRow.js (hunk 3 above)
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: Refactored. Main purpose - on Delete action fill empty indicator1, indicator2 and subfield fields in appropriate row with asterisk

File: src/components/MARCTable/MARCTableRow.js (hunk 4 above)
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: fixed

File: src/utils/formValidators.js (hunk 1 above)
Comment (IvanKryzhanovskyi) [reply src/utils/formValidators.js]: fixed

File: src/components/MARCTable/MARCTableRow.js [hunk 5]
```diff
@@ -85,14 +93,28 @@ export const MARCTableRow = ({
...
+    name: 'indicator2',
+  }, {
+    value: subfieldValue || '',
+    name: 'subfield',
+  }];
+
   const validateTag = useCallback(
     value => {
-      if (actionValue === MAPPING_DETAILS_ACTIONS.ADD) {
+      if (actionValue === MAPPING_DETAILS_ACTIONS.ADD || actionValue === MAPPING_DETAILS_ACTIONS.DELETE) {
```
[the remark this answers is not in the document]
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: fixed

File: src/components/MARCTable/MARCTableRow.js [hunk 6]
```diff
@@ -113,7 +139,11 @@ export const MARCTableRow = ({
   const validateIndicator2 = useCallback(
     value => {
       if (actionValue === MAPPING_DETAILS_ACTIONS.ADD) {
```
[the remark this answers is not in the document]
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: fixed

File: src/components/MARCTable/MARCTableRow.js [hunk 7]
```diff
@@ -113,7 +139,11 @@ export const MARCTableRow = ({
...
-        return validateAlphanumericValue(value) || validateMarcIndicatorField(fieldValue, indicator1Value, value);
+        return validateAlphanumericOrAllowedValue(value) || validateMarcIndicatorField(fieldValue, indicator1Value, value);
+      }
+
+      if (actionValue === MAPPING_DETAILS_ACTIONS.DELETE) {
```
[the remark this answers is not in the document]
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: fixed

File: src/components/MARCTable/MARCTableRow.js [hunk 8]
```diff
@@ -123,7 +153,11 @@ export const MARCTableRow = ({
   const validateSubfield = useCallback(
     value => {
       if (actionValue === MAPPING_DETAILS_ACTIONS.ADD) {
-        return validateAlphanumericValue(value) || validateSubfieldField(fieldValue)(value);
+        return validateAlphanumericOrAllowedValue(value) || validateSubfieldField(fieldValue)(value);
+      }
+
+      if (actionValue === MAPPING_DETAILS_ACTIONS.DELETE) {
```
[the remark this answers is not in the document]
Comment (IvanKryzhanovskyi) [reply src/components/MARCTable/MARCTableRow.js]: fixed

File: src/components/MARCTable/MARCTable.js [hunk 9]
```diff
@@ -165,6 +187,7 @@ export const MARCTable = ({
         onAddSubfieldRow={addSubfieldRow}
         onRemoveSubfieldRow={removeSubfieldRow}
         onRemoveSubfieldRows={removeSubfieldRows}
+        onRemoveActionSelect={fillEmptyFieldOnDeleteAction}
```
Comment (mariia-aloshyna) [inline src/components/MARCTable/MARCTable.js:190]:
```suggestion
        onDeleteActionSelect={fillEmptyFieldOnDeleteAction}
```

File: src/components/MARCTable/MARCTable.js [hunk 10]
```diff
@@ -132,6 +132,28 @@ export const MARCTable = ({
...
+            ...rowToUpdate.field,
+            ...(!isSubfieldField && { [item.name]: item.value }),
+            ...(isSubfieldField && { subfields: [{ subfield: item.value }] }),
+          },
+        };
+
+        onChange(updatedRows);
+      });
+  };
+
```
Comment (mariia-aloshyna) [inline src/components/MARCTable/MARCTable.js:156]:
```suggestion
  const fillEmptyFieldsWithValue = (order, fieldNames, valueToFill) => {
    const updatedRows = [...rows];
    const rowToUpdateIndex = rows.findIndex(row => row.order === order);
    const rowToUpdate = JSON.parse(JSON.stringify(updatedRows[rowToUpdateIndex]));

    fieldNames.forEach(path => {
      const fieldValue= get(rowToUpdate, path, '');

      if (!fieldValue) {
        set(rowToUpdate, path, valueToFill);
      }
    });

    updatedRows[rowToUpdateIndex] = rowToUpdate;

    onChange(updatedRows);
  };
```

**Pull request merged**